### ISSUE TITLE
Fabric.all_gather should support concatenation

### ISSUE THREAD

**Issue by cemde** [Lightning-AI/lightning#18607]
### Description & Motivation

When using DistributedDataParallel, the "Fabric.all_gather" option is useful to collect tensors from all devices. `x = fabric.all_gather(x)` will return a Tensor of dimensions `[input.shape[0], fabric.world_size, input.shape[1:]]`, which is natural for a lot of further processing.

### Pitch

It would be useful, if there was a flag `fabric.all_gather(x, op=OP, dim=1)` where `OP` can be `"stack"` and `"cat"`. For example, it would be possible to directly output `[batch_size_per_gpu * fabric_world_size, ...]`, which is more natural for processing the input to compute metrics etc. I assume it is also faster to directly concatenate instead of stacking first and then re-ordering.

### Alternatives

Alterantive is to keep doing this
```python
            if STRATEGY == "ddp":
                y_hat = fabric.all_gather(y_hat)
                y = fabric.all_gather(y)
                y_hat, y = y_hat.flatten(0,1), y.flatten(0,1)
```

**Comment by awaelchli** [comment]
Hey @cemde 

Thanks for the suggestion!
The output format you see is the result of `torch.distributed.all_gather` which we call under the hood. The backend in torch doesn't support any way to configure this, so even if Fabric supported this then it wouldn't be any faster since Fabric would have to do the reshaping. So I'm pretty sure that we can't deliver on this part of improving the efficiency unfortunately. 

We can still gather opinions and discuss whether the convenience of specifying the dimension would make sense though :)

**Comment by carmocca** [comment]
I agree with Adrian. This is very specific to the backend used. For instance, I faced difficulties in the past using XLA's broadcast with different number of tensors and/or different shapes. This often seems to be an underlying assumption and the implementation needs to be checked for each backend.
